interface/forms/phq9/phq9_javasrc.js

```diff
@@ -0,0 +1,140 @@
+var total_digits = null;  //element to hold the displayed digits
+var view_start = true; // edit is starting up, we need to read in the previous scores from DB
+var str_score_analysis = [" " + xl("No depressive disorder"), " " + xl("Minimal depressive disorder"), " " + xl("Moderate depressive disorder"), " " + xl("Moderately severe depressive disorder"), " " + xl("Severe depressive disorder")];
+
+/**
+ * manage question 10 - it is only be displayed when the score is not zero. It's answer is not
+ * included in the score.
+ * If necessary create and display question 10, needs to be done on startup or when score changes
+ * Default value of it's menu is either 'please select answer' or previous value selected
+ * If score has gone to zero then remove question 10 from the display if necessary
+ * @param  int value - index of previous answer to question 10
+ *                     'undef' (please select answer) if first time q8 being displayed
+ * @return undefined
+ */
+function manage_question_10(value)
+{
+    if ((phq9_score > 0) && q10_gone) {
+        question = document.createElement("i");  // create the node to hold question 10
+        question.class = "label";
+        var menue = document.createElement("select");  // create 'select' element, to hold the menue
+        // set some of the parameters
+        menue.name = "difficulty";
+        menue.onchange = "record_score_q10(my_form.difficulty.value);";
+        menue.length = 5;
+        create_q10(question, menue); // populate question 10 and menue - do in main page as it requires php
+        // set the default value - if new it's 'select answer', else it's previous value
+        if (value == "undef") {
+            menue.options[4].defaultSelected = true;
+        } else {
+            // else we can use value as an index
+            menue.options[Number(value)].defaultSelected = true;
+        }
+        //    display the question and  menue in the reserved place
+        place = document.getElementById("q10_place");
+        place.parentNode.appendChild(question, place);
+        place.parentNode.appendChild(menue, place);
+        q10_gone = false;
+    } else if (phq9_score == 0 && !q10_gone) {    //take question 10 off the displayed form
+        document.my_form.difficulty.remove();
+        question.remove();
+        q10_gone = true;
+    }
+    // nothing to do as
+    //   score > 0 but the question is already there -
+    // or score == 0 and this is at startup
+}
+
+// function update_score - display new total score - check if question 8 should be displayed
+// @param int index  question being answered, is 'undef' if we simply want to display the score, e.g. on startup
+// @param int new_score is 'undef' if it's from clicking 'please select an answer' in a new form - treat as zero.
+// @return true|false
+function update_score(index, new_score)
+{
+  //index is the number of the question, score it's score
+    var score = new_score;
```

Comment:
can you make this
```
    var score = parseInt(new_score);
```
and then make the suggested changes to use `js_escape()` instead of `attr()` to keep the use of this function consistent?

Reply:
hi @stephenwaite - i'm afraid the use of parseInt at this point does not fix the problem introduced by using js_escape. If i use parseInt when i obtain the value, i.e. where i've just used js_escape, then I get a numeric value.This works as the numeric score (the sum of the individual scores for each question) is calculated by adding the number obtained each time. by the time update_score is called if new_score is already a string of digits then  parseInt is simply converting the whole of this string into a, rather large, integer. 

e.g. if new_score is '134' instead of 8, then  parseInt   returns the integer 134 - a score off the scale (which is why the explanation is ' Severe depressive disorder' - the highest it can be.

 It would help me enormously if I understood why it's important to use js_escape(), rather than, say, text() -  can you explain or point me at documentation,  

meanwhile, if attr() is depricated here, it seems the choice is 
 var i = <?php echo js_escape($obj['interest_score']); ?>; //the value from last time
                            phq9_score += parseInt(i);
or
 var i = <?php echo text($obj['interest_score']); ?>; //the value from last time
                            phq9_score += i;
which is more straight forward.  
both work correctly. 

The alternative would be to rewrite the logic so it copes with strings of numbers - but that is much more work
thanks

Reply:
hi @ruthkonyn , [escaping functions](https://github.com/openemr/openemr/blob/master/library/htmlspecialchars.inc.php)

Reply:
thanks @stephenwaite - i had already found this file. what i don't understand is why i would want to get a json string rather than an integer straight off, which is what i need. What is the problem with my using text() at this point? If it's important i'm happy to change all these instances, as indicated in the first code snippet above, i'm just very curious as to why, so i can write better code in the future. Put it down to my inexperience with writing web applications... thanks

Reply:
looks like we'll need a new function like
```
/**
 * Escape a javascript literal for maths.
 */
function js_escape_num($text)
{
    return json_encode($text, JSON_NUMERIC_CHECK);
}

Reply:
> function js_escape_num($text)
> {
>     return json_encode($text, JSON_NUMERIC_CHECK);
> }

Wait! Why? 
From what I see all values being scored are strings so json_encode is doing its job. Otherwise if a score was truly an integer then json_encode would handle and pass the integer.
The problem I see is you're trying to do math with an intermix of string and integers in javascript where javascript would determine to use string concatenation based on if any argument in expression is a string.

You need to do the correct math in your scoring function as shouldn't need new escaping routines

Reply:
hi @sjpadgett - my original code used a series of lines like 
 ```
 var i = <?php echo text($obj['interest_score']); ?>; //the value from last time
                            phq9_score +=  i;
                            
```
to build a score from a number of answers on a form. It was suggested to use attr() rather than text, then js_escape() - conversations with stephen and brady above and on earlier pr at #5562. then stephen suggested the solution above, which works nicely. 

yes, using js_escape generates string concatanation, whereas text() or attr() seem to generate integer values.

the form answers are strings, but they are associated with numeric values. 

i guess i can accept a concatenation of numbers as strings and then write code to deal with this in the scoring function - it's just a bit more work. 

please could you agree with @stephenwaite and @bradymiller what you'd like me to do, and let me know, then we can get this form submitted, many thanks